### ISSUE TITLE
Add `StatefulDataloader` to remainder of recipes

### ISSUE THREAD

**Issue by joecummings** [pytorch/torchtune#2439]
**Goal:** Propagate use of `StatefulDataloader` to all recipes in torchtune.

**Why?** In order to successfully set up step-based checkpointing, we need to be able to track and store the dataloader state.

**How?** I went through an added `StatefulDataloader` to 5 recipes in #2410 and #2431. You should use these as a guide on how to properly add support and test the changes. At a high level the changes are:
- Import statefuldataloader from torchdata, optionally import statefuldistributedsampler if modifying a distributed recipe
- In `save_checkpoint` grab the current state dict from the dataloader
- In `setup` utilize the dataloader state dict and load it into the newly instantiated dataloader, then advance the iterator
- Move the section that breaks the iteration mid epoch based on `max_steps_per_epoch` to the bottom of the training loop. We don't want to advance the iterator here before breaking out.
- If there is a test and you are modifying a single device recipe, modify the loss values to match the new random state
- IF THERE IS NOT A TEST, you should confirm that you can kick off a training run and that you can resume from training. WandB logs are proof

**What are the remaining recipes that need this change?**
- [ ] knowledge_distillation_distributed
- [ ] knowledge_distillation_single_device
- [ ] qat_lora_finetune_distributed
- [ ] qat_distributed
- [ ] ppo_full_finetune_single_device
- [ ] full_dpo_distributed
- [ ] lora_dpo_distributed
- [ ] lora_dpo_single_device

**Comment by krammnic** [comment]
Will take `lora_dpo_single_device` But, if I can be a little bit greedy, I can do any other.

**Comment by krammnic** [comment]
@joecummings This can be closed now!
